src/modules/multiblade/MBCaenInstrument.cpp

```diff
@@ -0,0 +1,204 @@
+      monitorprod.produce(DataBuffer, Timestamp);
+    };
+    histfb.set_callback(ProduceHist);
+    histograms = Hists(std::max(ncass * nwires, ncass * nstrips), 65535);
+    histfb = HistogramSerializer(histograms.needed_buffer_size(), "multiblade");
+    //
+
+}
+
+
+// Moved from MBCaenBase to better support unit testing
+bool MBCaenInstrument::parsePacket(char * data, int length,  EV42Serializer & ev42ser) {
+
+  int res = parser.parse(data, length);
+
+  counters.ReadoutsErrorBytes += parser.Stats.error_bytes;
+  counters.ReadoutsErrorVersion += parser.Stats.error_version;
+  counters.ReadoutsSeqErrors += parser.Stats.seq_errors;
+
+  if (res < 0) {
+    return false;
+  }
+
+  XTRACE(DATA, DEB, "Received %d readouts from digitizer %d",
+         parser.MBHeader->numElements, parser.MBHeader->digitizerID);
+
+  counters.ReadoutsCount += parser.MBHeader->numElements;
+
+  uint64_t efu_time = 1000000000LU * (uint64_t)time(NULL); // ns since 1970
+  ev42ser.pulseTime(efu_time);
+
+  if (dumpfile) {
+    dumpfile->push(parser.readouts);
+  }
+
+  auto cassette = MultibladeConfig.Mappings->cassette(parser.MBHeader->digitizerID);
+  if (cassette < 0) {
+    XTRACE(DATA, WAR, "Invalid digitizerId: %d", parser.MBHeader->digitizerID);
+    counters.PacketBadDigitizer++;
+    return false;
+  }
+
+  FixJumpsAndSort(cassette, parser.readouts);
+  builders[cassette].flush();
+
+  return true;
+}
+
+
+// New EF algorithm - Needed to sort readouts in time
+bool compareByTime(const Readout &a, const Readout &b) {
+  return a.local_time < b.local_time;
+}
+
+// New EF algorithm - buffers data according to time and sorts before
+// processing
+void MBCaenInstrument::FixJumpsAndSort(int cassette, std::vector<Readout> &vec) {
+  int64_t Gap{43'000'000};
+  int64_t PrevTime{0xffffffffff};
+  std::vector<Readout> temp;
+
+  for (auto &Readout : vec) {
+    int64_t Time = (uint64_t)(Readout.local_time * MultibladeConfig.TimeTickNS);
+
+    if ((PrevTime - Time) < Gap) {
+      temp.push_back(Readout);
+    } else {
+      XTRACE(CLUSTER, DEB, "Wrap: %4d, Time: %lld, PrevTime: %lld, diff %lld",
+             counters.ReadoutsTimerWraps, Time, PrevTime, (PrevTime - Time));
+      counters.ReadoutsTimerWraps++;
+      std::sort(temp.begin(), temp.end(), compareByTime);
+      LoadAndProcessReadouts(cassette, temp);
+
+      temp.clear();
+      temp.push_back(Readout);
+    }
+    PrevTime = Time;
+  }
+  LoadAndProcessReadouts(cassette, temp);
+}
+
+//
+void MBCaenInstrument::LoadAndProcessReadouts(int cassette, std::vector<Readout> &vec) {
+  for (auto &dp : vec) {
+    if (not mbgeom.isValidCh(dp.channel)) {
+      counters.ReadoutsInvalidChannel++;
+      continue;
+    }
+
+    if (dp.adc > MultibladeConfig.max_valid_adc) {
+      counters.ReadoutsInvalidAdc++;
+      continue;
+    }
+
+    uint8_t plane = mbgeom.getPlane(dp.channel);
+
+    #ifdef MB_MONITOR_CHANNEL
+    if ((dp.digitizer == 137) and (dp.channel == 62) and (plane == 0)) {
```

Comment:
What are these specific numbers?

Reply:
Good catch will be replaced with defines - this is no longer in use but kept for backwards compatibility